go/enclave/db/storage.go

```diff
@@ -515,37 +517,105 @@ func (s *storageImpl) loadLogs(requestingAccount *gethcommon.Address, whereCondi
 			Topics: []gethcommon.Hash{},
 		}
 		var t0, t1, t2, t3, t4 sql2.NullString
-		err := rows.Scan(&t0, &t1, &t2, &t3, &t4, &l.Data, &l.BlockHash, &l.BlockNumber, &l.TxHash, &l.TxIndex, &l.Index, &l.Address)
+		err = rows.Scan(&t0, &t1, &t2, &t3, &t4, &l.Data, &l.BlockHash, &l.BlockNumber, &l.TxHash, &l.TxIndex, &l.Index, &l.Address)
 		if err != nil {
 			return nil, fmt.Errorf("could not load log entry from db: %w", err)
 		}
-		if t0.Valid {
-			l.Topics = append(l.Topics, hash(t0))
+
+		for _, topic := range []sql2.NullString{t0, t1, t2, t3, t4} {
+			if topic.Valid {
+				l.Topics = append(l.Topics, hashString(topic))
+			}
+		}
+
+		result = append(result, &l)
+	}
+
+	return result, rows.Close()
+}
+
+func (s *storageImpl) DebugGetLogs(txHash common.TxHash) ([]*tracers.DebugLogs, error) {
+	var queryParams []any
+
+	query := `select distinct 
+    			relAddress1, relAddress2, relAddress3, relAddress4,
+    			lifecycleEvent,
+    			topic0, topic1, topic2, topic3, topic4, 
+    			datablob, blockHash, blockNumber, txHash, txIdx, logIdx, address 
+				from events 
+				where txHash = ?`
+
+	queryParams = append(queryParams, txHash.Bytes())
+
+	result := make([]*tracers.DebugLogs, 0)
+
+	rows, err := s.db.GetSQLDB().Query(query, queryParams...)
+	if err != nil {
+		return nil, err
+	}
+
+	for rows.Next() {
+		l := tracers.DebugLogs{
+			Log: types.Log{
+				Topics: []gethcommon.Hash{},
+			},
+			LifecycleEvent: false,
 		}
-		if t1.Valid {
-			l.Topics = append(l.Topics, hash(t1))
+
+		var t0, t1, t2, t3, t4 sql2.NullString
+		var relAddress1, relAddress2, relAddress3, relAddress4 sql2.NullByte
+		err = rows.Scan(
+			&relAddress1,
+			&relAddress2,
+			&relAddress3,
+			&relAddress4,
+			&l.LifecycleEvent,
+			&t0, &t1, &t2, &t3, &t4,
+			&l.Data,
+			&l.BlockHash,
+			&l.BlockNumber,
+			&l.TxHash,
+			&l.TxIndex,
+			&l.Index,
+			&l.Address,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could not load log entry from db: %w", err)
 		}
-		if t2.Valid {
-			l.Topics = append(l.Topics, hash(t2))
+
+		for _, topic := range []sql2.NullString{t0, t1, t2, t3, t4} {
+			if topic.Valid {
+				l.Topics = append(l.Topics, hashString(topic))
+			}
 		}
-		if t3.Valid {
-			l.Topics = append(l.Topics, hash(t3))
+
+		mappedHashes := map[int]*gethcommon.Hash{
```

Comment:
This map of pointers is quite confusing to me and I think there might be an off-by-one error in the loop.

If I'm reading it correctly, I think it would be clearer and shorter to just duplicate like:

```
l.RelAddress1 = hashBytes(relAddress1)
l.RelAddress2 = hashBytes(relAddress2)
l.RelAddress3 = hashBytes(relAddress3)
l.RelAddress4 = hashBytes(relAddress4)
```

since we don't expect this list to grow ever.

Reply:
Yeah good point. It was off by one. Moved the validation to the hashbytes method 👏